### PR TITLE
End to end planning

### PR DESCRIPTION
**Pull request by halil93ibrahim**
Hello,
As discussed, I am creating this PR to target the develop branch instead of master.

### CODE REVIEW

**Comment by ad-daniel** [comment]
FYI, github is having issues this morning, some errors in the CI are related to that

**Comment by halil93ibrahim** [comment]
Hi, since the issue is related to the pre-trained model (basically some attribute names have changed with different versions of Gym), I replaced saved_model with a placeholder while trying. It seems the problem is solved.
I will include true saved_model and notify you then.

**Comment by halil93ibrahim** [comment]
Hi, I have also updated the documentation and I think it is ready for review now.

File: .github/workflows/tests_suite.yml
Comment (ad-daniel) [inline .github/workflows/tests_suite.yml:113]:
```suggestion
```

Comment (ad-daniel) [reply .github/workflows/tests_suite.yml:113]: It seems to me you're not using webots afterall, so we can skip this step

File: .github/workflows/tests_suite.yml
Comment (ad-daniel) [inline .github/workflows/tests_suite.yml:78]: Will this tool only work if installed from github? The `test-docker` job at least should work as well, no? (in the same workflow file, towards the end of the file)

File: .github/workflows/tests_suite_develop.yml
Comment (ad-daniel) [reply .github/workflows/tests_suite_develop.yml:113]: Ditto

File: .github/workflows/tests_suite_develop.yml
Comment (ad-daniel) [inline .github/workflows/tests_suite_develop.yml:77]: ditto.

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:1]: for consistency with the rest, the file should be named: `end-to-end-planning.md`

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:91]:
```suggestion
```

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:78]: I believe the infer method wasn't documented.

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:96]:
```suggestion
For the installation of Ardupilot instructions are available [here](https://github.com/ArduPilot/ardupilot).
```

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:99]: This phrase sounds a bit odd. The utility downloads the zip and the user should replace the contents?

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:100]:
```suggestion
In order to run Ardupilot in Webots 2021a, controller codes should be replaced. 
```

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:105]: probably better to provide the instructions, like:

```shell
sudo apt install this-and-that
```

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:111]:
```suggestion
- Start the Webots and open the provided world file.
The simulation time should stop at first time step and wait for Ardupilot software to run.
```

Comment (ad-daniel) [reply docs/reference/end_to_end_planning.md:111]: one phrase per line

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:113]:
```suggestion
- Run following script from Ardupilot's directory: `./libraries/SITL/examples/Webots/dronePlus.sh` which starts software in the loop execution of the Ardupilot software.
```

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:130]: To simplify things we prefer shortened imports.
Basically you need to create a `__init__.py` file as done [here](https://github.com/opendr-eu/opendr/blob/master/src/opendr/perception/face_recognition/__init__.py) and then you can import simply by doing:
`from opendr.planning.end_to_end_planning import EndToEndPlanningRLLearner`

File: docs/reference/end_to_end_planning.md
Comment (ad-daniel) [inline docs/reference/end_to_end_planning.md:142]: import can be simplified here as well

File: src/opendr/planning/end_to_end_planning/README.md
Comment (ad-daniel) [inline src/opendr/planning/end_to_end_planning/README.md:11]: same comments made for the docs

File: tests/sources/tools/planning/end_to_end_planning/test_end_to_end_planning.py
Comment (ad-daniel) [inline tests/sources/tools/planning/end_to_end_planning/test_end_to_end_planning.py:20]: These imports can be simplified as well after the change

File: src/opendr/planning/end_to_end_planning/src/take_off.cpp
Comment (ad-daniel) [inline src/opendr/planning/end_to_end_planning/src/take_off.cpp:62]: still needed?

File: src/opendr/planning/end_to_end_planning/envs/agi_env.py
Comment (ad-daniel) [inline src/opendr/planning/end_to_end_planning/envs/agi_env.py:35]: Some of the files only mention OpenDR as rights owner (for instance `take_off.cpp) whereas other files, like this, mention both OpenDR and a MIT license. Is it intentional? Does that mean some of the work was performed outside of the OpenDR context?

File: .github/workflows/tests_suite.yml
Comment (halil93ibrahim) [reply .github/workflows/tests_suite.yml:78]: Done

File: .github/workflows/tests_suite_develop.yml
Comment (halil93ibrahim) [reply .github/workflows/tests_suite_develop.yml:77]: done

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: done

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: Yes, it was forgotten. I have added.

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: Done

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: Rephrased.

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: Done

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: I have provided the link for instructions.

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: done

File: docs/reference/end_to_end_planning.md
Comment (halil93ibrahim) [reply docs/reference/end_to_end_planning.md]: Simplified

File: src/opendr/planning/end_to_end_planning/src/take_off.cpp
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/src/take_off.cpp]: Deleted

File: src/opendr/planning/end_to_end_planning/envs/agi_env.py
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/envs/agi_env.py]: I think I have seen it from another tool. I have deleted MIT licenses.

**Comment by halil93ibrahim** [comment]
Hi Daniel, thank you for the comments. I think I have fixed them.
Also, I have uploaded an updated 'ardupilot.zip' file to the cloud. Can you upload it to the FTP?

**Comment by ad-daniel** [comment]
I don't think I have access to the FTP, I believe @nikolaid77 handled the other uploads

**Comment by Pavlos-Tosidis** [comment]
> 

Updated file is now uploaded.

**Comment by halil93ibrahim** [comment]
Hi, can you help me to solve the error below? I have created __init__ files under planning and end_to_end_planning folders.

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3.8/unittest/__main__.py", line 18, in <module>
    main(module=None)
  File "/usr/lib/python3.8/unittest/main.py", line 100, in __init__
    self.parseArgs(argv)
  File "/usr/lib/python3.8/unittest/main.py", line [12](https://github.com/opendr-eu/opendr/runs/5540846924?check_suite_focus=true#step:3:12)4, in parseArgs
    self._do_discovery(argv[2:])
  File "/usr/lib/python3.8/unittest/main.py", line 244, in _do_discovery
    self.createTests(from_discovery=True, Loader=Loader)
  File "/usr/lib/python3.8/unittest/main.py", line [15](https://github.com/opendr-eu/opendr/runs/5540846924?check_suite_focus=true#step:3:15)4, in createTests
    self.test = loader.discover(self.start, self.pattern, self.top)
  File "/usr/lib/python3.8/unittest/loader.py", line 346, in discover
    raise ImportError('Start directory is not importable: %r' % start_dir)
ImportError: Start directory is not importable: 'tests/sources/tools/planning/end_to_end_planning'

**Comment by ad-daniel** [comment]
Should probably wait for https://github.com/opendr-eu/opendr/pull/215 to be merged since some dependencies are changed and I think a submodule also changed. I'm not sure, but develop might be currently broken, I'll take a closer look when it's merged. In principle no, shouldn't be necessary to add those init files, or it triggers the entire test suite instead of just testing the sources.

**Comment by halil93ibrahim** [comment]
Hi @ad-daniel , it still gets the error: `ImportError: Start directory is not importable: tests/sources/tools/planning/end_to_end_planning'` when testing dockers.
Why do you think that happens?

**Comment by ad-daniel** [comment]
Basically the issue was that when setting the `test release` label it builds the docker image from the develop branch in which, currently, your tool is not yet included hence when trying to run the e2e test, it doesn't find one. I've quickly tested to see if it would work by forcing it to build the image from this branch and it appears to work, so it will be ok once we merge the PR.

I think I wasn't clear concerning the documentation, what I meant to say is that this PR should have the main doc, and when this one is merged, in the other PR you can mention the performance evaluation.

Now that the cuda PR was merged, this PR should also adapt to the changes introduced [there](https://github.com/opendr-eu/opendr/pull/215/files), both on the learner and test side concerning device/TEST_DEVICE. 

Lastly, we should mention the addition of this tool in the changelog.

**Comment by halil93ibrahim** [comment]
So, I have mentioned this tool as a new feature in the changelog. Do you think the description is ok?

File: CHANGELOG.md
Comment (ad-daniel) [inline CHANGELOG.md:7]:
```suggestion
    - Added end-to-end planning tool ([#223](https://github.com/opendr-eu/opendr/pull/223)).
```

**Review by ad-daniel — CHANGES_REQUESTED** [review]
I've adjusted it, only the other issue mentioned in the previous message missing now and it's fine for me.

**Comment by halil93ibrahim** [comment]
Do you mean the device/TEST_DEVICE part? Can you elaborate on what I should do?

**Review by ad-daniel — DISMISSED** [review]
Looks good to me, thank you.

PS: I noticed you moved the entire docs file to the other PR, what I meant in that PR is you should only complete it once this one was merged. To keep things simple can you move it back here? (it's fine if you leave the performance evaluation included)

**Comment by halil93ibrahim** [comment]
I have deleted the docs file in the other PR now. The file in this PR already includes the performance evaluation.

**Review by passalis — CHANGES_REQUESTED** [review]
Overall this seems fine! However, there are some minor inconsistencies with the rest of the toolkit. More specifically, `fit()` and `eval()` are expected to return dictionaries (e.g., see https://github.com/opendr-eu/opendr/blob/master/src/opendr/perception/pose_estimation/lightweight_open_pose/lightweight_open_pose_learner.py#L585), while `save/load()` should follow OpenDR format, which is essentially a data wrapper for any other format (e.g., see https://github.com/opendr-eu/opendr/blob/master/src/opendr/perception/pose_estimation/lightweight_open_pose/lightweight_open_pose_learner.py#L679)

**Comment by halil93ibrahim** [comment]
I have updated `fit()` and `eval()` to return dictionaries. However, for `save/load()` I am following a similar approach with MobileRL: 
https://github.com/opendr-eu/opendr/blob/a7bd1d1d9b1ec53ffa9e25f282cc7524cb27d2c0/src/opendr/control/mobile_manipulation/mobile_manipulation_learner.py#L209
We both depend on stable-baselines3 for RL implementation, so it is the stable-baselines3 format used. Do you think it needs to be changed?

**Comment by passalis** [comment]
I think it is ok! Thank you!

**Review by passalis — APPROVED** [review]

**Comment by halil93ibrahim** [comment]
Thanks! It says somehow I dismissed ad-daniel's approval. Did I do something wrong?

**Comment by ad-daniel** [comment]
No it's fine, it's because you made a change therefore it needs to be re-approved. I'll do so when the issues brought up by Nikos have been addressed

**Review by ad-daniel — APPROVED** [review]
Thank you

**Pull request merged**
